File: dailyQuestion/2020-07/07-01/python/solution_window.py
def findLength(A: [int], B: [int]) -> int:
    def maxLength(addA: int, addB: int, length: int) -> int:
        ret = k = 0
        for i in range(length):
            if A[addA + i] == B[addB + i]:
                k += 1
                ret = max(ret, k)
            else:
                k = 0
        return ret

    n, m = len(A), len(B)
    ret = 0
    for i in range(n):
        length = min(m, n - i)
        ret = max(ret, maxLength(i, 0, length))
    for i in range(m):
        length = min(n, m - i)
        ret = max(ret, maxLength(0, i, length))
    return ret

if __name__ == "__main__" :

    A = [1,2,3,2,1]
    B = [3,2,1,4,7]

    a = A[0:1] == B[2:3]
    result = findLength(A,B)
    print(result)

File: dailyQuestion/2020-07/07-01/golang/solution_window.go
package main

import "fmt"

func findLength(A []int, B []int) int {
	m := len(A)
	n := len(B)
	result := 0
	var maxLength func(addA, addB, length int) int
	maxLength = func(addA, addB, length int) int {
		ret := 0
		k := 0
		for i := 0; i < length; i++ {
			if A[addA+i] == B[addB+i] {
				k++
				ret = max(ret, k)
			} else {
				k = 0
			}
		}
		return ret
	}

	for i := 0; i < m; i++ {
		length := min(n, m-i)
		result = max(result, maxLength(i, 0, length))
	}
	for i := 0; i < n; i++ {
		length := min(m, n-i)
		result = max(result, maxLength(0, i, length))
	}

	return result
}

func max(x, y int) int {
	if x > y {
		return x
	}
	return y
}

func min(x, y int) int {
	if x < y {
		return x
	}
	return y
}

func main() {
	A := []int{1, 2, 3, 2, 1}
	B := []int{3, 2, 1, 4, 7}
	result := findLength(A, B)
	fmt.Println(result)
}
